### PR TITLE
[backend] new: add get gpu counts API

### PR DESCRIPTION
**Pull request by under-chaos**

### CODE REVIEW

File: ymir/backend/src/pymir-controller/controller/invoker/invoker_cmd_gpu.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+class GPUInvoker(BaseMirControllerInvoker):
+    def pre_invoke(self) -> backend_pb2.GeneralResp:
+        return checker.check_request(request=self._request, prerequisites=[checker.Prerequisites.CHECK_USER_ID],)
+
+    def invoke(self) -> backend_pb2.GeneralResp:
+        if self._request.req_type != backend_pb2.CMD_GPU_GET:
+            raise RuntimeError("Mismatched req_type")
+
+        free_gpus = gpu_utils.GPUInfo.get_available_gpus()
```
Comment (phoenix-xhuang) [inline ymir/backend/src/pymir-controller/controller/invoker/invoker_cmd_gpu.py]: available_gpus =

File: ymir/backend/src/pymir-controller/controller/invoker/invoker_cmd_gpu.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from controller.invoker.invoker_cmd_base import BaseMirControllerInvoker
+from controller.utils import checker, utils, gpu_utils, code
+from proto import backend_pb2
+
+
+class GPUInvoker(BaseMirControllerInvoker):
```
Comment (phoenix-xhuang) [inline ymir/backend/src/pymir-controller/controller/invoker/invoker_cmd_gpu.py]: GPUInfoInvoker

File: ymir/backend/src/common/proto/backend.proto [hunk 3]
```diff
@@ -89,6 +89,7 @@ enum RequestType {
     CMD_TERMINATE = 14;
     CMD_LABLE_TASK_TERMINATE = 15 [deprecated = true];
     CMD_PULL_IMAGE = 16;
+    CMD_GPU_GET = 17;
```
Comment (phoenix-xhuang) [inline ymir/backend/src/common/proto/backend.proto]: CMD_GPU_INFO_GET

Comment (phoenix-xhuang) [reply ymir/backend/src/common/proto/backend.proto]: 在GeneralResp增加一个int available_gpu_counts，用于返回结果。现在GeneralResp里没有可复用的字段

File: ymir/backend/src/pymir-controller/controller/invoker/invoker_cmd_gpu.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    def pre_invoke(self) -> backend_pb2.GeneralResp:
+        return checker.check_request(request=self._request, prerequisites=[checker.Prerequisites.CHECK_USER_ID],)
+
+    def invoke(self) -> backend_pb2.GeneralResp:
+        if self._request.req_type != backend_pb2.CMD_GPU_GET:
+            raise RuntimeError("Mismatched req_type")
+
+        free_gpus = gpu_utils.GPUInfo.get_available_gpus()
+
+        return utils.make_general_response(code.ResCode.CTR_OK, str(len(free_gpus)))
```
Comment (phoenix-xhuang) [inline ymir/backend/src/pymir-controller/controller/invoker/invoker_cmd_gpu.py]: response = backend_pb2.GeneralResp()
response.code = code.ResCode.CTR_OK
response.available_gpu_counts = len(available_gpus)

Comment (under-chaos) [reply ymir/backend/src/pymir-controller/controller/invoker/invoker_cmd_gpu.py]: Done

File: ymir/backend/src/pymir-controller/controller/utils/invoker_mapping.py [hunk 5]
```diff
@@ -14,7 +14,8 @@
     invoker_cmd_label_get,
     invoker_cmd_label_add,
     invoker_cmd_terminate,
-    invoker_cmd_pull_image
+    invoker_cmd_pull_image,
+    invoker_cmd_gpu
```
Comment (phoenix-xhuang) [inline ymir/backend/src/pymir-controller/controller/utils/invoker_mapping.py]: 把这里和RequestTypeToInvoker里面的顺序调整下吧

Comment (under-chaos) [reply ymir/backend/src/pymir-controller/controller/utils/invoker_mapping.py]: Done

File: ymir/backend/src/pymir-controller/controller/utils/invoker_mapping.py [hunk 6]
```diff
@@ -14,7 +15,7 @@
     invoker_cmd_label_get,
     invoker_cmd_label_add,
     invoker_cmd_terminate,
-    invoker_cmd_pull_image
+    invoker_cmd_pull_image,
```
Comment (phoenix-xhuang) [inline ymir/backend/src/pymir-controller/controller/utils/invoker_mapping.py]: invoker_cmd_branch_checkout,
    invoker_cmd_branch_commit,
    invoker_cmd_branch_create,
    invoker_cmd_branch_delete,
    invoker_cmd_branch_list,
    invoker_cmd_filter,
    invoker_cmd_gpu_info,
    invoker_cmd_inference,
    invoker_cmd_init,
    invoker_cmd_label_add,
    invoker_cmd_label_get,
    invoker_cmd_log,
    invoker_cmd_merge,
    invoker_cmd_pull_image,
    invoker_cmd_task_info,
    invoker_cmd_terminate,
    invoker_task_factory,

File: ymir/backend/src/pymir-controller/controller/utils/invoker_mapping.py [hunk 7]
```diff
@@ -37,4 +38,5 @@
     backend_pb2.CMD_LABEL_GET: invoker_cmd_label_get.LabelGetInvoker,
     backend_pb2.CMD_TERMINATE: invoker_cmd_terminate.CMDTerminateInvoker,
     backend_pb2.CMD_PULL_IMAGE: invoker_cmd_pull_image.ImageHandler,
+    backend_pb2.CMD_GPU_INFO_GET: invoker_cmd_gpu_info.GPUInfoInvoker
```
Comment (phoenix-xhuang) [inline ymir/backend/src/pymir-controller/controller/utils/invoker_mapping.py]: backend_pb2.CMD_BRANCH_CHECKOUT: invoker_cmd_branch_checkout.BranchCheckoutInvoker,
    backend_pb2.CMD_BRANCH_CREATE: invoker_cmd_branch_create.BranchCreateInvoker,
    backend_pb2.CMD_BRANCH_DEL: invoker_cmd_branch_delete.BranchDeleteInvoker,
    backend_pb2.CMD_BRANCH_LIST: invoker_cmd_branch_list.BranchListInvoker,
    backend_pb2.CMD_COMMIT: invoker_cmd_branch_commit.BranchCommitInvoker,
    backend_pb2.CMD_FILTER: invoker_cmd_filter.FilterBranchInvoker,
    backend_pb2.CMD_GPU_INFO_GET: invoker_cmd_gpu_info.GPUInfoInvoker,
    backend_pb2.CMD_INFERENCE: invoker_cmd_inference.InferenceCMDInvoker,
    backend_pb2.CMD_INIT: invoker_cmd_init.InitInvoker,
    backend_pb2.CMD_LABEL_ADD: invoker_cmd_label_add.LabelAddInvoker,
    backend_pb2.CMD_LABEL_GET: invoker_cmd_label_get.LabelGetInvoker,
    backend_pb2.CMD_LOG: invoker_cmd_log.LogInvoker,
    backend_pb2.CMD_MERGE: invoker_cmd_merge.MergeInvoker,
    backend_pb2.CMD_PULL_IMAGE: invoker_cmd_pull_image.ImageHandler,
    backend_pb2.CMD_TERMINATE: invoker_cmd_terminate.CMDTerminateInvoker,
    backend_pb2.REPO_CREATE: invoker_cmd_init.InitInvoker,
    backend_pb2.TASK_CREATE: invoker_task_factory.CreateTaskInvokerFactory,
    backend_pb2.TASK_INFO: invoker_cmd_task_info.GetTaskInfoInvoker,

**Review by phoenix-xhuang — APPROVED** [review]

**Review by IJtLJZ8Rm4Yr — APPROVED** [review]

**Pull request merged**